pkg/extract/redhat/csaf/csaf.go

```diff
@@ -680,216 +689,330 @@ func walkVulnerabilities(vulns []csaf.Vulnerability, pids []csaf.ProductID) (map
+				tag := calculateTag(major, pidsPerMajor, status)
+				segment := segmentTypes.Segment{
+					Ecosystem: ecosystemTypes.Ecosystem(fmt.Sprintf("%s:%s", ecosystemTypes.EcosystemTypeRedHat, major)),
+					Tag:       tag,
+				}
+
+				adv.Segments = append(adv.Segments, segment)
+
+				for cveID, ass := range vassGroup {
+					if ass.Status != status {
+						continue
+					}
+					baseV, found := baseVulnerabilities[cveID]
+					if !found {
+						return nil, nil, nil, errors.Errorf("no base vulnerability found for %s", cveID)
+					}
+
+					sevs, err := buildSeverities(ass.Severity)
 					if err != nil {
-						return nil, errors.Wrapf(err, "parse cvss2")
+						return nil, nil, nil, errors.Wrap(err, "build severities")
 					}
-					ss = append(ss, severityTypes.Severity{
-						Type:   severityTypes.SeverityTypeCVSSv2,
-						Source: "secalert@redhat.com",
-						CVSSv2: v2,
+					baseV.Severity = sevs
+					baseV.Sort()
+
+					index := slices.IndexFunc(vs, func(v vulnerabilityTypes.Vulnerability) bool {
+						return vulnerabilityContentTypes.Compare(v.Content, baseV) == 0
 					})
+					if index < 0 {
+						vs = append(vs, vulnerabilityTypes.Vulnerability{
+							Content:  baseV,
+							Segments: []segmentTypes.Segment{segment},
+						})
+					} else {
+						vs[index].Segments = append(vs[index].Segments, segment)
+					}
 				}
-				if ass.severity.cvss3 != "" {
-					switch {
-					case strings.HasPrefix(ass.severity.cvss3, "CVSS:3.0"):
-						v30, err := cvssV30Types.Parse(ass.severity.cvss3)
-						if err != nil {
-							return nil, errors.Wrap(err, "parse cvss3")
+
+				var cs1 []criterionTypes.Criterion
+				for _, pid := range pidsPerMajor {
+					for _, p := range pm[pid] {
+						if p.arch == "src" {
+							continue
 						}
-						ss = append(ss, severityTypes.Severity{
-							Type:    severityTypes.SeverityTypeCVSSv30,
-							Source:  "secalert@redhat.com",
-							CVSSv30: v30,
-						})
-					case strings.HasPrefix(ass.severity.cvss3, "CVSS:3.1"):
-						v31, err := cvssV31Types.Parse(ass.severity.cvss3)
+
+						vc, err := buildVersionCriterion(p, status)
 						if err != nil {
-							return nil, errors.Wrap(err, "parse cvss3")
+							return nil, nil, nil, errors.Wrap(err, "build version criterion")
 						}
-						ss = append(ss, severityTypes.Severity{
-							Type:    severityTypes.SeverityTypeCVSSv31,
-							Source:  "secalert@redhat.com",
-							CVSSv31: v31,
+						if vc == nil {
+							continue
+						}
+
+						index := slices.IndexFunc(cs1, func(c criterionTypes.Criterion) bool {
+							// compare by identity except architectures
+							return c.Version.Vulnerable == vc.Vulnerable &&
+								fixstatusTypes.Compare(*c.Version.FixStatus, *vc.FixStatus) == 0 &&
+								c.Version.Package.Binary.Name == vc.Package.Binary.Name &&
+								slices.Compare(c.Version.Package.Binary.Repositories, vc.Package.Binary.Repositories) == 0 &&
+								affectedTypes.Compare(*c.Version.Affected, *vc.Affected) == 0
```

Comment:
If you are doing a lot of comparisons and hoping they are equal, it may be better to use cmp.Or.
```suggestion
							return cmp.Or(
								func() int {
									switch {
									case c.Version.Vulnerable && !vc.Vulnerable:
										return 1
									case !c.Version.Vulnerable && vc.Vulnerable:
										return -1
									default:
										return 0
									}
								}(),
								fixstatusTypes.Compare(*c.Version.FixStatus, *vc.FixStatus),
								cmp.Compare(c.Version.Package.Binary.Name, vc.Package.Binary.Name),
								slices.Compare(c.Version.Package.Binary.Repositories, vc.Package.Binary.Repositories),
								affectedTypes.Compare(*c.Version.Affected, *vc.Affected),
							) == 0
```

---

> Or returns the first of its arguments that is not equal to the zero value. If no argument is non-zero, it returns the zero value.

https://pkg.go.dev/cmp#Or

Reply:
However, it may be easier to run it if you organize arch and repository in advance.